plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1236,39 +1236,59 @@ public void executeRemoveOrphanFiles(ConnectorSession session, IcebergTableExecu
     {
-        Set<String> validDataFilePaths = stream(table.snapshots())
-                .map(Snapshot::snapshotId)
-                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
-                .map(fileScanTask -> fileName(fileScanTask.file().path().toString()))
-                .collect(toImmutableSet());
-        Set<String> validDeleteFilePaths = stream(table.snapshots())
-                .map(Snapshot::snapshotId)
-                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
-                .flatMap(fileScanTask -> fileScanTask.deletes().stream().map(file -> fileName(file.path().toString())))
-                .collect(Collectors.toUnmodifiableSet());
-        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, union(validDataFilePaths, validDeleteFilePaths), "/data");
-    }
+        Set<String> validMetadataFilePaths = new HashSet<>();
+        ImmutableSet.Builder<String> validDataFilePaths = ImmutableSet.builder();
+        ImmutableSet.Builder<String> validDeleteFilePaths = ImmutableSet.builder();
 
-    private void removeOrphanMetadataFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
-    {
-        ImmutableSet<String> manifests = stream(table.snapshots())
-                .flatMap(snapshot -> snapshot.allManifests(table.io()).stream())
-                .map(ManifestFile::path)
-                .collect(toImmutableSet());
-        List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
-        List<String> otherMetadataFiles = concat(
-                metadataFileLocations(table, false).stream(),
-                Stream.of(versionHintLocation(table)))
-                .collect(toImmutableList());
-        Set<String> validMetadataFiles = concat(manifests.stream(), manifestLists.stream(), otherMetadataFiles.stream())
-                .map(IcebergMetadata::fileName)
-                .collect(toImmutableSet());
-        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validMetadataFiles, "metadata");
+        for (Snapshot snapshot : table.snapshots()) {
+            if (snapshot.manifestListLocation() != null) {
+                validMetadataFilePaths.add(snapshot.manifestListLocation());
+            }
+
+            for (ManifestFile manifest : snapshot.allManifests(table.io())) {
+                if (validMetadataFilePaths.contains(manifest.path())) {
+                    // Already read this manifest
+                    continue;
+                }
+
+                validMetadataFilePaths.add(manifest.path());
+                if (manifest.content() == ManifestContent.DATA) {
+                    try (ManifestReader<DataFile> manifestReader = ManifestFiles.read(manifest, table.io());
+                        CloseableIterator<DataFile> dataFiles = manifestReader.iterator()) {
```

Comment:
do you need to close this closeable iterator?

Reply:
I take that back, ManifestReader is a CloseableGroup so it handles it when it closes